dynamiqs/integrators/apis/mcsolve.py

```diff
@@ -0,0 +1,272 @@
+    MCSolveEulerIntegrator,
+    MCSolveIntegrator,
+    MCSolveKvaerno3Integrator,
+    MCSolveKvaerno5Integrator,
+    MCSolveTsit5Integrator,
+)
+
+
+def mcsolve(
+    H: QArrayLike | TimeArray,
+    jump_ops: list[QArrayLike | TimeArray],
+    psi0: QArrayLike,
+    tsave: ArrayLike,
+    *,
+    keys: ArrayLike = jax.random.split(jax.random.key(31), num=10),  # noqa: B008
+    exp_ops: list[QArrayLike] | None = None,
+    solver: Solver = Tsit5(),  # noqa: B008
+    root_finder: AbstractRootFinder | None = optx.Newton(1e-5, 1e-5, optx.rms_norm),  # noqa: B008
+    gradient: Gradient | None = None,
+    options: Options = Options(),  # noqa: B008
+) -> MCSolveResult:
+    r"""Perform Monte-Carlo evolution, unraveling the master equation.
+
+    We follow the algorithm outlined in Abdelhafez et al. to efficiently perform
+    Monte-Carlo sampling. First the no-jump trajectory is computed for a state vector
+    $\ket{\psi(t)}$ at time $t$, starting from an initial state $\ket{\psi_0}$,
+    according to the Schrödinger equation with non-Hermitian Hamiltonian ($\hbar=1$)
+    $$
+        \frac{\dd\ket{\psi(t)}}{\dt}
+        = -i [H(t) -\frac{i}{2} \sum_{k=1}^{N} L_{k}^{\dagger} (t)L_{k}(t)] \ket{\psi(t)},
+    $$
+    where $H(t)$ is the system's Hamiltonian at time $t$ and $\{L_k(t)\}$ is a
+    collection of jump operators at time $t$. We then extract the norm of the state
+    at the final time, and take this as a lower bound for random numbers sampled
+    for other trajectories such that they all experience at least one jump.
+
+    Quote: Time-dependent Hamiltonian or jump operators
+        If the Hamiltonian or the jump operators depend on time, they can be converted
+        to time-arrays using [`dq.constant()`](/python_api/time_array/constant.html),
+        [`dq.pwc()`](/python_api/time_array/pwc.html),
+        [`dq.modulated()`](/python_api/time_array/modulated.html), or
+        [`dq.timecallable()`](/python_api/time_array/timecallable.html).
+
+    Quote: Running multiple simulations concurrently
+        The Hamiltonian `H`, the jump operators `Ls` and the
+        initial state `psi0` can be batched to solve multiple monte-carlo equations
+        concurrently. All other arguments are common to every batch.
+
+    Args:
+        H _(qarray-like or time-array of shape (bH?, n, n))_: Hamiltonian.
+        Ls _(list of qarray-like or time-array, of shape (nL, n, n))_: List of
+            jump operators.
+        psi0 _(qarray-like of shape (bpsi?, n, 1))_: Initial state.
+        tsave _(array-like of shape (nt,))_: Times at which the states and expectation
+            values are saved. The equation is solved from `tsave[0]` to `tsave[-1]`, or
+            from `t0` to `tsave[-1]` if `t0` is specified in `options`.
+        keys _(Array of shape (ntraj,))_: Total number of jump trajectories to
+            simulate, not including the no-jump trajectory. Defaults to a list of keys
+            of length 10.
```

Comment:
Should state explicitely that the `keys` length corresponds to the number of trajectories sampled, similarly to https://github.com/dynamiqs/dynamiqs/blob/ef8cc5fe3e4dba6987313a621e9559b268c988d0/dynamiqs/integrators/apis/dsmesolve.py#L134-L136.

Reply:
Right: in fact the behavior here is a bit nonstandard and depends on `options.cartesian_batching`: here is my proposed change. Let me know if you think this api should be changed
```suggestion
        keys _(Array of shape (...ntraj,))_: PRNG keys used to sample the jump trajectories.
            With options.cartesian_batching=True, this array must be 1D and the number of
            elements defines the number of sampled trajectories, not including the no-jump
            trajectory. With options.cartesian_batching=False, the leading dimensions 
            of keys must be broadcastable with those of H, Ls, psi0, with the final
            dimension determining the number of jump trajectories.  
```